engine/baml-lib/jsonish/src/deserializer/coercer/match_string.rs

```diff
@@ -142,57 +142,76 @@ fn string_match_strategy<'c>(
+
+    // Look for substrings of valid values
+    for (variant, valid_names) in candidates {
+        for valid_name in valid_names {
+            for (start_idx, _) in value_str.match_indices(valid_name) {
+                let end_idx = start_idx + valid_name.len();
+                all_matches.push((start_idx, end_idx, valid_name, variant));
+            }
+        }
+    }
 
-    // Sort by max count, then min pos.
-    result.sort_by(|a, b| match a.0.cmp(&b.0) {
-        Ordering::Less => Ordering::Greater,
-        Ordering::Greater => Ordering::Less,
-        Ordering::Equal => a.1.cmp(&b.1),
+    // No substring match at all for any variant, early return.
+    if all_matches.is_empty() {
+        return None;
+    }
+
+    // Sort by position and length
+    all_matches.sort_by(|a, b| {
+        match a.0.cmp(&b.0) {
+            Ordering::Equal => b.1.cmp(&a.1), // Longer first
+            ordering => ordering, // Less or Greater stays the same
+        }
     });
 
-    // Filter for max count.
-    let max_count = result.first().map(|r| r.0).unwrap_or(0);
-    result.retain(|r| r.0 == max_count);
+    // Filter out overlapping matches
+    let mut filtered_matches = Vec::new();
+    let mut last_end = 0;
+
+    for current_match in all_matches {
+        if current_match.0 >= last_end {
+            // No overlap with previous match
+            last_end = current_match.1;
+            filtered_matches.push(current_match);
+        }
+    }
+
+    // Count occurrences of each variant in non-overlapping matches.
+    // (count, variant)
+    let mut variant_counts: Vec<(usize, &'c str)> = Vec::new();
+    for (_, _, _, variant) in &filtered_matches {
+        // Increment count if variant already exists
+        if let Some(entry) = variant_counts.iter_mut().find(|(_, v)| v == variant) {
+            entry.0 += 1;
+        } else {
+            // Add new variant
+            variant_counts.push((1, variant));
+        }
+    }
+
+    // Sort by count
+    variant_counts.sort_by(|a, b| b.0.cmp(&a.0));
 
-    // Return the best match if there is one.
-    if let Some((_, _, candidate)) = result.first() {
+    // Return the best match if there is one
+    if let Some((count, candidate)) = variant_counts.first() {
         flags.add_flag(Flag::SubstringMatch(value_str.into()));
 
-        // Add flag for multiple matches.
-        if result.len() > 1 {
-            flags.add_flag(Flag::StrMatchOneFromMany(
-                result
-                    .iter()
-                    .map(|(count, _, candidate)| ((*count) as usize, candidate.to_string()))
-                    .collect(),
-            ));
+        // Find all variants with the same count
+        let max_count = count;
+        let ties: Vec<_> = variant_counts
+            .iter()
+            .filter(|(c, _)| c == max_count)
+            .map(|(count, variant)| (*count as usize, variant.to_string()))
```

Comment:
Judging by the code [here](https://github.com/BoundaryML/baml/blob/23e590b0b2fdb51f80e7eced769baabd12b3be22/engine/baml-lib/jsonish/src/deserializer/deserialize_flags.rs#L178-L183) it looks like the first element of the tuple should be the index of the match, not its count:

```rust
Flag::StrMatchOneFromMany(values) => {
    write!(f, "Enum one from many: ")?;
    for (idx, value) in values {
        writeln!(f, "Item {}: {}", idx, value)?;
    }
}
```

I'm fixing this right now.

Reply:
I guess that code is wrong and `idx` is actually `count`, changed it in [this commit](https://github.com/BoundaryML/baml/pull/1098/commits/555fd242fd673e74d0f35c77db2bffd70ab8bdd9). Maybe we should add both the count and all the occurrences to every match so we can return a nicer error in [`deserialize_flags.rs`](https://github.com/BoundaryML/baml/blob/23e590b0b2fdb51f80e7eced769baabd12b3be22/engine/baml-lib/jsonish/src/deserializer/deserialize_flags.rs#L178-L183), but for now this PR looks good to me!